fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java

```diff
@@ -718,6 +723,68 @@ protected RuntimeException mapAsyncToSyncException(@Nonnull Throwable ex) {
         }
     }
 
+    /**
+     * Join a future, following the same logic that asyncToSync uses to validate that the operation isn't blocking
+     * in an asynchronous context.
+     *
+     * @param future the future to be completed
+     * @param <T> the type of the value produced by the future
+     * @return the result value
+     */
+    public <T> T join(CompletableFuture<T> future) {
+        checkIfBlockingInAsync(future);
+        return future.join();
+    }
+
+    /**
+     * Get a future, following the same logic that asyncToSync uses to validate that the operation isn't blocking
+     * in an asynchronous context.
+     *
+     * @param future the future to be completed
+     * @param <T> the type of the value produced by the future
+     * @return the result value
+     *
+     * @throws java.util.concurrent.CancellationException if the future was cancelled
+     * @throws ExecutionException if the future completed exceptionally
+     * @throws InterruptedException if the current thread was interrupted
+     */
+    public <T> T get(CompletableFuture<T> future) throws InterruptedException, ExecutionException {
+        checkIfBlockingInAsync(future);
+        return future.get();
+    }
+
+    @API(API.Status.INTERNAL)
+    public BlockingInAsyncDetection getBlockingInAsyncDetection() {
+        return blockingInAsyncDetectionSupplier.get();
+    }
+
+    private void checkIfBlockingInAsync(CompletableFuture<?> future) {
+        BlockingInAsyncDetection behavior = getBlockingInAsyncDetection();
+        if (behavior == BlockingInAsyncDetection.DISABLED) {
+            return;
+        }
+
+        final boolean isComplete = future.isDone();
+        if (isComplete && behavior.ignoreComplete()) {
+            return;
+        }
+
+        final StackTraceElement[] stack = Thread.currentThread().getStackTrace();
+        for (StackTraceElement stackElement : stack) {
+            if (stackElement.getClassName().startsWith(CompletableFuture.class.getName())) {
```

Comment:
One thing this doesn't handle (I don't think) is something like:

```java
@Nonnull
public CompletableFuture<Void> doSomethingAsync(@Nonnull FDBRecordStore store) {
    Message record = store.loadRecord(Tuple.from(1066L));
    return AsyncUtil.DONE;
}
```

This exact example is somewhat contrived as (hopefully) the user knows that loading a record requires doing I/O, but there are other instances that are somewhat more subtle.

I think there isn't a *general* way of solving this that doesn't have its own set of issues. I wonder if it's worth adding a check to see if `Async` is in the stack element? Or does that produce its own set of problems.